spec/ics-2-consensus-verification/README.md

```diff
@@ -0,0 +1,310 @@
+of how does verifiers, cstates, headers work & generated. I think we should move these
+to another paragraph, as the other types and functions are actually used by other icss*
+
+`Commit` is a `Header` generating function which takes the previous `Header` with the messages
+and returns the result.
+
+```go
+type Commit func(ConsensusState, [Message]) Header
+```
+
+##### Blockchain
+
+Defined as blockchain consensus algorithm which generates valid `Header`s.
+It generatess a unique list of headers starting from a genesis `ConsensusState` with arbitrary
+messages.
+
+`Blockchain` is defined as
+```go
+type Blockchain struct {
+  Verifier Verifier
+  Genesis ConsensusState
+  Commit Commit
+}
+```
+where
+  * `Verifier` is the lightclient verifier
+  * `Genesis` is the genesis `ConsensusState`
+  * `Commit` is the header generating function
+
+The headers generated from the `Blockchain` is expected to satisfy the followings:
+
+1. The `Header`s have no more than one direct child
+
+* Satisfied if: deterministic safety
+* Possible violation scenario: validator double signing, chain reorganization (Nakamoto consensus)
+
+2. The `Header`s eventually have at least one direct child
+
+* Satisfied if: liveness, light-client verifier continuity
+* Possible violation scenario: synchronised halt, incompatible hard fork
+
+3. The `Header`s are generated from the `Commit`, which ensures valid transition of the state
+
+* Satisfied if: correct block generation & state machine
+* Possible violation scenario: invariant break, validator cartel
+
+In these cases it is ensured that the `Verifier` works as intended.
+
+##### LightClient
+
+LightClient is defined as
+```go
+type LightClient struct {
+  Verifier Verifier
+  ConsensusState ConsensusState
+}
+```
+where
+  * `ConsensusState` is the root of trust providing the `VerifierBase`
+  * `Verifier` is the lightclient verification logic
+
+The exact type of each fields are depending on the type of the actual consensus logic.
+
+##### Functions
+
+* `Height :: Blockchain -> Header -> Uint` returns the position of the header in the
+  Blockchain's header list.
+
+#### Requirements
+
+Consensus verification requires the following accumulator primitives with datastructures and
+properties as defined in ICS23:
+
+* `AccumulatorRoot`
+
+#### Subprotocols
+
+##### Getter/Setter
+
+Identifiers are used to point the LightClients.
+
+```go
+function getLightClient(id :: Identifier) LightClient { /*impl specific*/ }
+
+function setLightClient(id :: Identifier, lc :: LightClient) { /*impl specific*/ }
+```
+
+##### Register
+The method that the identifiers are given to the lightclients(`newLightClientID`)
+is not defined by the protocol, rather depending on the implementation. Registering
+a new LightClient is expected to be permissionless.
+
+```go
+type RegisterLightClient LightClient
+
+function register(register :: RegisterLightClient) Identifier {
```

Comment:
We need to be clear that this is a datagram which can be submitted directly to the IBC handler